### ISSUE TITLE
Propagate type information in compiled template code to allow static analysis

### ISSUE THREAD

**Issue by MartinMystikJonas** [nette/latte#262]
I just finished first draft of script that uses PHPStan to satic analysis of compiled Latte templates in our projects.

One think I noticed is that {varType} or {templateType} macros does not propagate any type related information to compiled Latte templates so PHPStan cannot check it.

Is there any plan to address this?

I could try to prepare PR which will address this but I would like to disccuss what approach to take here.

One option would be to simply add relevant `/** @var type $variable` annotations after each extract of parameters. his is only way for 

Another IMHO better approach would be to generate known params as template properties and instead of addressign them as `$variable` use `$this->variable` if variable is known but I am not sure how difficult that change would be.

**Comment by lulco** [comment]
Hi @MartinMystikJonas, @dg

I am thinking about this too...

What about this:

```
/** @var int $a */
$a = $this->params['a'];

/** @var string $b */
$b = $this->params['b'];

/** @var Selection $c */
$c = $this->params['c'];

extract($this->params);

```
where $a, $b, $c annotations are from varType or properties from templateType and could be used for static analysis, and extract($this->params) just keep it working like before this change.

**Comment by jakubvojacek** [comment]
Subscribing, that'd be cool!

**Comment by jakubvojacek** [comment]
Anyone managed to come up with somewhat usable solution how to analyse lattes via phpstan? The `@var` annotation would certainly be helpful

**Comment by lulco** [comment]
@jakubvojacek I don't think so. `@var` annotations are just first part of it. I tried to add them to compiled templates and run phpstan on them but there are many other issues like filters etc.

**Comment by jakubvojacek** [comment]
too bad, that'd be killer feature :(

**Comment by MartinMystikJonas** [comment]
I have managed to solve it. It is just proof of concept at the moment and without var annotations it is missing important checks. But it works. At the momeny I simply use PHpStan stubs to define known filters but I have idea how that can by solved better.

**Comment by MartinMystikJonas** [comment]
If you are intetested I will share my implementation with you guys.

**Comment by jakubvojacek** [comment]
Please, that'd be perfect, I am sure many would appreciate that

**Comment by lulco** [comment]
@MartinMystikJonas I am interested too :)

**Comment by MartinMystikJonas** [comment]
Here you go: https://github.com/MetisFW/phpstan-latte

**Comment by jakubvojacek** [comment]
Thank you, managed to make it work. But seems like that without the var annotation it won't find much useful. Most of the errors we get are accessing non existing properties etc. Which the phpstan would have caught if he knew the type

**Comment by MartinMystikJonas** [comment]
Well that is exactly the reason I started this issue in the first place. It wont catch most common errors without propagation of type information to compiled template.

**Comment by MartinMystikJonas** [comment]
@jakubvojacek @lulco I created PR that will propagate type information into resulting PHP code. See: https://github.com/nette/latte/pull/275
